kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -1784,17 +1843,20 @@ protected void handleDeleteGroupsRequest(KafkaHeaderAndRequest deleteGroups,
                                              CompletableFuture<AbstractResponse> resultFuture) {
         checkArgument(deleteGroups.getRequest() instanceof DeleteGroupsRequest);
         DeleteGroupsRequest request = (DeleteGroupsRequest) deleteGroups.getRequest();
+        DeleteGroupsRequestData data = request.data();
 
         resultFuture.complete(KafkaResponseUtils.newDeleteGroups(
-                getGroupCoordinator().handleDeleteGroups(request.groups())
+                getGroupCoordinator().handleDeleteGroups(data.groupsNames())
         ));
     }
 
     @Override
     protected void handleSaslAuthenticate(KafkaHeaderAndRequest saslAuthenticate,
                                           CompletableFuture<AbstractResponse> resultFuture) {
-        resultFuture.complete(new SaslAuthenticateResponse(Errors.ILLEGAL_SASL_STATE,
-                "SaslAuthenticate request received after successful authentication"));
+        SaslAuthenticateResponseData data = new SaslAuthenticateResponseData();
+        data.setErrorCode(Errors.ILLEGAL_SASL_STATE.code());
+        data.setErrorMessage("SaslAuthenticate request received after successful authentication");
+        resultFuture.complete(new SaslAuthenticateResponse(data));
```

Comment:
```suggestion
        resultFuture.complete(new SaslAuthenticateResponse(
                new SaslAuthenticateResponseData()
                        .setErrorCode(Errors.ILLEGAL_SASL_STATE.code())
                        .setErrorMessage("SaslAuthenticate request received after successful authentication")));
```

I think Kafka's new data structure is designed for chained calls above. The *data* is mutable while the *response* that owns the data is immutable.